### PR TITLE
net: tcp: Solved mystery of seqno double-increment

### PR DESCRIPTION
**Pull request by pfalcon**
Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

### CODE REVIEW

**Comment by pfalcon** [comment]
I'm happy to report that the mystery of double-increment of seqno is resolved. It can traced to ddda50c214, which in turns comes from FNET's implementation. I didn't freshly recheck, but I believe nowhere TCP spec talks about decrementing seqno's, the whole idea is that they're monotonically increase. So, that another case of "original design" (in wikipedia terms) in FNET.

Now, I only briefly tested that this seems to work OK with echo_server. TODO:

1. For maintainers/TCP code contributors to review the changes speculatively.
2. Actually thoroughly test with different examples.

**Comment by pfalcon** [comment]
To remind, double-increment was introduce in d6b93b311f961896640fa873bc6cf7ae221b0e1f with a pledge for it to be resolved eventually.

**Review by tbursztyka — CHANGES_REQUESTED** [review]
Even for rfc, apply the code style properly. At least to see if tests still pass (right now it just stops at checkpatch)

File: subsys/net/ip/tcp.c [hunk 1]
```diff
@@ -477,7 +477,7 @@ int net_tcp_prepare_segment(struct net_tcp *tcp, u8_t flags,
 		 * Linux to detect active close on server side, and
 		 * to make Wireshark happy about sequence numbers.
 		 */
-		seq += 2;
+		seq += 1;
```
Comment (jukkar) [inline subsys/net/ip/tcp.c]: Remember to remove the comment above.

**Review by jukkar — COMMENTED** [review]
Excellent catch, just cleanup the PR and we are good to go. I did some testing and everything seemed to work just fine.

**Comment by jukkar** [comment]
@pfalcon can you cleanup the PR and re-send it

**Comment by pfalcon** [comment]
@jukkar: Thanks for looking into this and testing it. I don't haste with this one, because I'd like to take a chance and do more homework with this.

**Comment by pfalcon** [comment]
I'm glad I didn't haste with this and took time to review codepaths into net_tcp_prepare_segment(). The patch as was breaks active open (i.e. connect()). The general problem is that we have TCP handling split (and apparently duplicated or at least overlapping) between tcp.c and net_context.c. Anyway, net_context.c:send_syn*() calls into net_tcp_prepare_segment(), which previously incremented seqno on SYN. Will send updated patch.

**Comment by pfalcon** [comment]
Updated. @tbursztyka , please re-review (you have -1). @jukkar, please retest, in particular testing active open cases. Thanks.

**Review by jukkar — APPROVED** [review]
Tested this and it works nicely.

**Comment by pfalcon** [comment]
Thanks!

File: subsys/net/ip/tcp.c [hunk 2]
```diff
@@ -485,24 +485,14 @@ int net_tcp_prepare_segment(struct net_tcp *tcp, u8_t flags,
...
 		 */
 		flags |= NET_TCP_ACK;
-		/* FIXME: We apparently miss increment in another
-		 * transition of the state machine, so have to
-		 * adjust seq no by 2 here. This is required for
-		 * Linux to detect active close on server side, and
-		 * to make Wireshark happy about sequence numbers.
-		 */
-		seq += 2;
+		seq += 1;
```
Comment (lpereira) [inline subsys/net/ip/tcp.c]: Very minor nit, but this could be `seq++`.

**Review by lpereira — COMMENTED** [review]
This is something that bugged me while working on TCP.  Thanks for fixing this -- I've tried a few times but there were always something that broke.

The TCP implementation living in both net_context.c and tcp.c is also something that bugged me for a while.  It's something that I'd love to see done someday; should clean up a bunch of things.

File: subsys/net/ip/tcp.c (hunk 2 above)
Comment (pfalcon) [reply subsys/net/ip/tcp.c]: Fixed

**Comment by pfalcon** [comment]
> The TCP implementation living in both net_context.c and tcp.c is also something that bugged me for a while. It's something that I'd love to see done someday; should clean up a bunch of things.

Hope to get there too eventually ;-). So far, going thru smaller things e.g. https://github.com/zephyrproject-rtos/zephyr/pull/4443 , https://github.com/zephyrproject-rtos/zephyr/issues/4421

**Review by tbursztyka — APPROVED** [review]

**Pull request merged**